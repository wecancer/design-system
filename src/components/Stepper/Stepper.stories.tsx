import React from 'react'
import {Story} from '@storybook/react'
import Stepper, {Props} from '.'

const Template: Story<Props> = (args) => <Stepper {...args} />

const items = [
  {id: '1', label: 'Step 1'},
  {id: '2', label: 'Step 2'},
  {id: '3', label: 'Step 3'},
]

export const Default = Template.bind({})
Default.args = {items}

export default {
  title: 'Components/Stepper',
  component: Stepper,
}
