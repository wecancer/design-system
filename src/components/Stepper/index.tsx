import React from 'react'
import styled, {css} from 'styled-components'
import Icon from '../Icon'

const Wrapper = styled.div`
  ${({theme}) => css`
    width: 100%;
    display: flex;
    font-family: ${theme.font.familyRedesign};
  `}
`

const Step = styled.div<{isActived: boolean}>`
  ${({isActived, theme}) =>
    css`
      width: 32px;
      height: 32px;
      display: flex;
      align-items: center;
      justify-content: center;
      border-radius: 50%;
      font-weight: ${theme.font.weights.bold};
      z-index: 1;

      ${isActived
        ? css`
            color: ${theme.colors.bg};
            background-color: ${theme.colors.primary};
          `
        : css`
            color: ${theme.colors.label};
            background-color: ${theme.colors.line};
          `}
    `}
`

const Label = styled.p<{isActived: boolean}>`
  ${({theme, isActived}) => css`
    color: ${isActived ? theme.colors.titleActive : theme.colors.label};
    text-align: center;
  `}
`

const Cell = styled.div<{isLastItem: boolean; isCompleted: boolean}>`
  ${({theme, isLastItem, isCompleted}) => css`
    display: flex;
    flex: 1;
    flex-direction: column;
    align-items: center;
    position: relative;

    ${!isLastItem &&
    css`
      &::after {
        content: '';
        width: 100%;
        height: 2px;
        position: absolute;
        top: 16px;
        left: 50%;
        background-color: ${isCompleted
          ? theme.colors.primary
          : theme.colors.line};
      }
    `}
  `}
`

export type Props = {
  current: number
  items: {
    id: string
    label?: string
  }[]
}

const Stepper = ({items, current = 1}: Props): React.ReactElement | null => {
  if (!items.length) return null

  return (
    <Wrapper
      role="progressbar"
      aria-valuenow={current}
      aria-valuemin={0}
      aria-valuemax={items.length}
    >
      {items.map((step, index) => {
        const isCompleted = current - 1 > index
        const isLastItem = items.length - 1 === index
        const isActived = index + 1 <= current
        const isChecked = index + 1 < current

        return (
          <Cell key={step.id} isLastItem={isLastItem} isCompleted={isCompleted}>
            <Step isActived={isActived}>
              {isChecked ? <Icon name="check" /> : index + 1}
            </Step>
            {step.label && <Label isActived={isActived}>{step.label}</Label>}
          </Cell>
        )
      })}
    </Wrapper>
  )
}

export default Stepper
