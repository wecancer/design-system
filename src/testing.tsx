import React from 'react'
import {render, RenderResult} from '@testing-library/react'
import WecancerProvider from './WecancerProvider'

type Props = {
  children: React.ReactNode
}

const AllTheProviders = ({children}: Props) => {
  return <WecancerProvider>{children}</WecancerProvider>
}

// eslint-disable-next-line
const appRender = (ui: React.ReactElement, options?: any): RenderResult =>
  render(ui, {wrapper: AllTheProviders, ...options})

// re-export everything
export * from '@testing-library/react'

// override render method
export {appRender as render}
