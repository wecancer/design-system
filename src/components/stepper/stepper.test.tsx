import Stepper from '.'
import {render} from '../../testing'

describe('<Stepper />', () => {
  it('should render snapshot correctly', () => {
    const {container} = render(
      <>
        <Stepper items={[]} current={1} />
        <Stepper items={[{id: '1'}, {id: '2'}]} current={1} />
        <Stepper
          current={0}
          items={[
            {label: 'item 1', id: '1'},
            {label: 'item 2', id: '2'},
            {label: 'item 3', id: '3'},
          ]}
        />
        <Stepper
          current={1}
          items={[
            {label: 'item 1', id: '1'},
            {label: 'item 2', id: '2'},
            {label: 'item 3', id: '3'},
          ]}
        />
        <Stepper
          current={2}
          items={[
            {label: 'item 1', id: '1'},
            {label: 'item 2', id: '2'},
            {label: 'item 3', id: '3'},
          ]}
        />
        <Stepper
          current={3}
          items={[
            {label: 'item 1', id: '1'},
            {label: 'item 2', id: '2'},
            {label: 'item 3', id: '3'},
          ]}
        />
        <Stepper
          current={4}
          items={[
            {label: 'item 1', id: '1'},
            {label: 'item 2', id: '2'},
            {label: 'item 3', id: '3'},
          ]}
        />
      </>,
    )
    expect(container.firstChild).toMatchSnapshot()
  })

  it('should render aria properties correctly', () => {
    const {getByRole} = render(
      <Stepper
        current={2}
        items={[
          {label: 'item 1', id: '1'},
          {label: 'item 2', id: '2'},
          {label: 'item 3', id: '3'},
        ]}
      />,
    )

    expect(getByRole('progressbar').getAttribute('aria-valuenow')).toBe('2')
    expect(getByRole('progressbar').getAttribute('aria-valuemin')).toBe('0')
    expect(getByRole('progressbar').getAttribute('aria-valuemax')).toBe('3')
  })
})
