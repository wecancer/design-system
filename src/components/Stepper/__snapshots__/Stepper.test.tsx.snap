// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Stepper /> should render snapshot correctly 1`] = `
.c0 {
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-family: Poppins,sans-serif;
}

.c2 {
  width: 32px;
  height: 32px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 50%;
  font-weight: 700;
  z-index: 1;
  color: #F7F7FC;
  background-color: #0080EB;
}

.c4 {
  width: 32px;
  height: 32px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border-radius: 50%;
  font-weight: 700;
  z-index: 1;
  color: #6E7191;
  background-color: #D9DBE9;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

.c1::after {
  content: '';
  width: 100%;
  height: 2px;
  position: absolute;
  top: 16px;
  left: 50%;
  background-color: #D9DBE9;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
}

<div
  aria-valuemax="2"
  aria-valuemin="0"
  aria-valuenow="1"
  class="c0"
  role="progressbar"
>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      1
    </div>
  </div>
  <div
    class="c3"
  >
    <div
      class="c4"
    >
      2
    </div>
  </div>
</div>
`;
